Goodpratices, Technical Tools and New Technical Keys Steps (#260)
PrestaShop/docs: src/content/1.7/modules/sell/techvalidation-keysteps.md

---
title: Technical validation - Key steps
---

# Technical Validation : Key steps

> Have you ever wondered how the PrestaShop Addons Team treats the validation ?

Reviewer: Not sure if `treats` is the correct translation 🤔

"performs" ? "considers" ? "carries out" ?

Reviewer 2: Well, I would write _Have you ever wondered how the PrestaShop Addons Team proceeds with the validation?_, what do you think?

We are going to show you how to win time so you can pretty much "validate" your own module before submitting it on the Marketplace Addons.
First you might want to read the [Technical Tools]({{<ref "1.7/modules/sell/technical-tools" >}}) and the [Good Practices]({{<ref "1.7/modules/creation/good-practices" >}}). 

### 1. The Basics

> As most of you know, there are 3 types of submittions : New, Update min, Update maj.

Reviewer: ```suggestion
As you probably know, there are 3 types of submissions : New, Minor update, Major update.
```

We do not handle them with the same standards:

- We are very strict and meticulous when it comes to New modules.

> - For major updates, we have the exact same verifications than the New modules.

Reviewer: ```suggestion
- For major updates, we have the exact same verifications that the New modules.
```

> - For minor updates, we verify the validator and we also have a Diff Tool to only see the modifications.

Reviewer: ```suggestion
- For minor updates, we rely on the validator and we also have a Diff Tool to only see the modifications.
```

> Let's now give you more details about how we handle shit.

Reviewer: 🤣I dont think "shit" will be accepted by the master of wordings @LouiseBonnard

Author: Yeah i'm very sorry. I was going too fast to create an introduction so I could focus on the main content. I intended to rewrite that but totally forgot.

The danger of not having time to start and end it in a single day :(

> ### 2. Treatment of New and Major updates

Reviewer: Let's use the same replacement that for previous treat. Processing looks good here IMO

What's a "New" Module ? 
Actually, it's a Zip. Your module is the product and since you can submit several zips per Product, let's use the right words.

Simple: The very first time you submit a zip for your module to the marketplace. It's a "New" zip.
Please, don't even try to submit your module if you didn't ever use the [Validator](https://validator.prestashop.com/).

> That tool is very powerful and useful and it will make everbody win a lot of time.

Reviewer: ```suggestion
That tool is very powerful and useful and it will make everybody win a lot of time.
```

I think we should rather explain than anyway, the module will go through the Validator processing, so it's useless to skip it, it only will make the validation process longer.

> After fixing every feedback from the Validator, here's a list of extra points we verify: 

Reviewer: ```suggestion
After meeting each requirement provided by the Validator, here's a list of extra points we check: 
```

* The presence of DROP/ALTER of PrestaShop tables.
    * It's highly forbidden to apply any changes on PrestaShop Core tables. It's very very dangerous and we don't allow any risk.

* Except for [Payment Modules](https://github.com/PrestaShop/paymentexample/blob/master/paymentexample.php#L150), the use of iframes is STRICTLY FORBIDDEN.

> * Every hook. If any on them is empty, we decline the zip.

Reviewer 2: ```suggestion
* Every hook. If any of them is empty, we decline the zip.
```

>     * There a other things we decline like loading a JS file in the whole backoffice when it's not necessary. Make sure to only load what you need, when you need it.

Reviewer 2: ```suggestion
    * There are other things we decline like loading a JS file in the whole back office when it is unnecessary. Make sure to only load what you need, when you need it.
```

> * Missing index.php files in your directories ? We decline.

Reviewer: Do we explain (maybe in another part of the doc) why these index.php files are required ? I guess yes, maybe we can provide the link in order to remind people of the importance of these files.

Author: Yes it's explained in another article. I'm adding a link for explanations

    * Note that we have [a tool](https://github.com/jmcollin/autoindex) to help you fix that easily.

> * Modules have to have a certain structure when it comes to files. 

Reviewer: ```suggestion
* Modules are required to follow a specific folder convention. Files must be in the right folder.
```

>     * For example JS and CSS files must be inside views/. The Validator gives us all that. Make sure to respect that as much as possible !

Reviewer: ```suggestion
    * For example JS and CSS files must be inside views/. The Validator will check this. Make sure to respect that as much as possible !
```

> * We verify every SQL request to make sure you did cast your variables. Use (int) or pSQL() accordingly. 

Reviewer: ```suggestion
* We examine every SQL request to make sure you did cast your variables. Use (int) or pSQL() accordingly. 
```

* If you have php files to handle ajax or external calls, make sure to secure that file.
    * In order to do that, we invite you to create a unique token during the module's install and use it during the call verification.

* The use of serialize() is forbidden.
    * But you can use base64_encode. However we only allow it when it comes to communicate with an external service. 

* Do not load any external content.
    * Either javascript, css, image, zip, whatever, it's not allowed. We even saw some modules downloading another module. The zip you send on PrestaShop Addons must be totally self-sufficient. 

**Attention** :

> We do have extra rules for Payment Modules which require tons of security.

Reviewer: ```suggestion
We do have extra rules for Payment Modules as this type of modules require higher security.
```
"Tons of damage !"
